### PR TITLE
refactor(connector): migrate `anyhow::Error` to `ConnectorError` newtype

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Follow-up of #15086.

- Migrate all functions returning `anyhow::Result` to `ConnectorResult`.
 https://github.com/risingwavelabs/risingwave/blob/de0233c0d15cba6e79614f1e177c82ebc4bf188b/src/connector/src/error.rs#L24-L61
  - Provide default contexts for errors from external connector crates.
  - Provide ad-hoc contexts at different call sites for general error types like `ParseIntError`.
- Style improvements for using `context` and `bail` to wrap and construct errors.

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15042?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #14950** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14950?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #14953** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14953?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #14956** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14956?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #15031** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15031?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #15042** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15042?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #14954** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14954?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15042?utm_source=stack-comment).

File: src/connector/src/common.rs [hunk 1]
```diff
@@ -386,23 +385,30 @@ pub struct PulsarOauthCommon {
...
             match url.scheme() {
                 "s3" => {
                     let credentials = load_file_descriptor_from_s3(&url, aws_auth_props).await?;
-                    let mut f = NamedTempFile::new()?;
-                    f.write_all(&credentials)?;
-                    f.as_file().sync_all()?;
-                    temp_file = Some(f);
+                    temp_file = Some(
+                        create_credential_temp_file(&credentials)
+                            .context("failed to create temp file for pulsar credentials")?,
```
Comment (xxchan) [inline src/connector/src/common.rs:410]: 😄

File: src/connector/src/parser/json_parser.rs [hunk 2]
```diff
@@ -43,15 +44,16 @@ pub struct JsonAccessBuilder {
...
         if payload.is_empty() {
             self.value = Some("{}".into());
         } else {
             self.value = Some(payload);
         }
         let value = simd_json::to_borrowed_value(
             &mut self.value.as_mut().unwrap()[self.payload_start_idx..],
-        )?;
+        )
+        .context("failed to parse json payload")?;
```
Comment (xxchan) [inline src/connector/src/parser/json_parser.rs:56]: What about printing the payload here in the ctx?

Comment (xxchan) [reply src/connector/src/parser/json_parser.rs:56]: Oh, it seems we can't print it without clone

**Review by xxchan — APPROVED** [review]
LGTM!
At first I thought maybe it's not necessary to replace all `anyhow::Error` to the newtype (although this is easiest from the refactoring's point of view), but only need to change public APIs. But I quite like the added `.context()`s. This feels like a dynamic version of `thiserror` and thus I think we can use it more widely. 😄

**Comment by BugenZhao** [comment]
### Merge activity

* **Feb 23, 4:24 AM EST**: [Graphite](https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15042) failed to merge this pull request due to **This repository has GitHub's merge queue enabled, which is currently incompatible with Graphite**. Try your merge again, or report a bug if you see this consistently.

**Pull request merged**